Application Layer Dashboard

Dhruv-J opened the pull request:
This PR adds an Application Layer Dashboard to include visualizations for layer
seven Network Policies and HTTP values within the flows table. The changes
include the following:

1. A new dashboard named Application Layer Visibility, which contains two
panels, a Chord Panel to illustrate layer seven Network Policies, and a table
panel that shows the HTTP values for each flow.

2. Since the default grafana table plugin isn't able to handle nested JSON
values and aggregate them in a clean manner, this PR introduces a new plugin
called http-table-plugin that leverages the material-ui library to create a
simple table that includes the source and destination IPs and ports, with
expandable rows that show multiple HTTP value sets if a flow contains them.
This plugin is created using the updated @grafana/create-plugin tool, which
contains a .config directory for all plugin build and run configuration files.

3. Testing for the application layer dashboard and unit testing for the
http-table-plugin.

This PR builds upon #417, which adds additional fields for L7 visibility.

Dhruv-J commented:
will update docs with following images:
<img width="2695" alt="Screenshot 2023-10-04 at 4 14 41 PM" src="https://github.com/antrea-io/theia/assets/10016630/31d10982-102b-4158-9a14-35e649170769">
<img width="1453" alt="Screenshot 2023-10-04 at 4 14 52 PM" src="https://github.com/antrea-io/theia/assets/10016630/73720f39-c634-437b-a935-658eb458c20e">
<img width="2693" alt="Screenshot 2023-10-04 at 4 16 59 PM" src="https://github.com/antrea-io/theia/assets/10016630/068a1f7a-b804-4203-b906-03674d40e2d4">
<img width="1458" alt="Screenshot 2023-10-04 at 4 17 05 PM" src="https://github.com/antrea-io/theia/assets/10016630/1421138e-252d-480d-b8db-ad1f53673c0a">

Dhruv-J commented:
![Screenshot 2023-10-16 at 11 20 46 AM](https://github.com/antrea-io/theia/assets/10016630/fa4502a5-bfba-436b-ad4b-4de2934a7dba)
![Screenshot 2023-10-16 at 11 20 55 AM](https://github.com/antrea-io/theia/assets/10016630/e658177c-2921-4cce-8a25-7bfcbb813068)

tushartathgur commented on build/charts/theia/provisioning/dashboards/application_layer_dashboard.json:
```diff
@@ -0,0 +1,119 @@
+  "uid": "tCTkERMSk",
+  "version": 1,
+  "weekStart": ""
+}
```
add a blank line

tushartathgur commented on build/charts/theia/provisioning/dashboards/homepage.json:
```diff
@@ -576,7 +576,7 @@
             "id": 27,
             "options": {
-                "content": "The following seven dashboards are pre-built and are recommended for network flow\nvisualization.\n\n<span style=\"color:rgb(184, 119, 217)\">Flow Records Dashboard</span> displays the flow records being captured in the selected\ntime range.\n\n<span style=\"color:rgb(184, 119, 217)\">Pod-to-Pod Flows Dashboard</span> shows cumulative bytes and throughput of Pod-to-Pod ...
+                "content": "The following eight dashboards are pre-built and are recommended for network flow\nvisualization.\n\n<span style=\"color:rgb(184, 119, 217)\">Flow Records Dashboard</span> displays the flow records being captured in the selected\ntime range.\n\n<span style=\"color:rgb(184, 119, 217)\">Pod-to-Pod Flows Dashboard</span> shows cumulative bytes and throughput of Pod-to-Pod ...
```
Application Visibility Dashboard</span> displays cumulative bytes and throughput for L7 NetworkPolicies\nas well as corresponding HTTP data.

tushartathgur commented on build/charts/theia/provisioning/dashboards/network_topology_dashboard.json:
```diff
@@ -32,15 +32,56 @@
+            }
+          },
+          "queryType": "sql",
+          "rawSql": "SELECT sourcePodName, sourcePodLabels, sourcePodNamespace, sourceNodeName, destinationPodName, destinationPodLabels, destinationNodeName, destinationServicePortName, octetDeltaCount FROM flows\nWHERE sourcePodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')\nAND destinationPodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')\nAND ( d ...
```
Shouldn't it be based on the l7ProtocolName instead of httpvals?

yuntanghsu replied:
Do we want to exclude the ns` local-path-storage`, which is created in the kind cluster by default?

tushartathgur commented on build/charts/theia/provisioning/dashboards/network_topology_dashboard.json:
```diff
@@ -60,11 +101,11 @@
           },
           "queryType": "sql",
-          "rawSql": "SELECT sourcePodName, sourcePodLabels, sourcePodNamespace, sourceNodeName, destinationPodName, destinationPodLabels, destinationNodeName, destinationServicePortName, octetDeltaCount FROM flows\nWHERE sourcePodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')\nAND destinationPodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')\nAND des ...
+          "rawSql": "SELECT sourcePodName, sourcePodLabels, sourcePodNamespace, sourceNodeName, sourceIP, sourceTransportPort, destinationPodName, destinationPodLabels, destinationNodeName, destinationServicePortName, destinationIP, destinationTransportPort, octetDeltaCount, httpVals FROM flows\nWHERE sourcePodNamespace NOT IN ('kube-system', 'flow-visibility', 'flow-aggregator')\nAND destination ...
```
ditto

tushartathgur commented on build/charts/theia/provisioning/dashboards/network_topology_dashboard.json:
```diff
@@ -102,6 +143,6 @@
+  "version": 3,
   "weekStart": ""
-}
+}
```
Blank line

yuntanghsu replied:
Solved?

tushartathgur commented on build/yamls/flow-visibility.yml:
```diff
@@ -2770,7 +2928,7 @@ data:
                 "id": 27,
                 "options": {
-                    "content": "The following seven dashboards are pre-built and are recommended for network flow\nvisualization.\n\n<span style=\"color:rgb(184, 119, 217)\">Flow Records Dashboard</span> displays the flow records being captured in the selected\ntime range.\n\n<span style=\"color:rgb(184, 119, 217)\">Pod-to-Pod Flows Dashboard</span> shows cumulative bytes and throughput of Pod-to ...
+                    "content": "The following eight dashboards are pre-built and are recommended for network flow\nvisualization.\n\n<span style=\"color:rgb(184, 119, 217)\">Flow Records Dashboard</span> displays the flow records being captured in the selected\ntime range.\n\n<span style=\"color:rgb(184, 119, 217)\">Pod-to-Pod Flows Dashboard</span> shows cumulative bytes and throughput of Pod-to ...
```
same as above

tushartathgur commented on plugins/grafana-custom-plugins/grafana-dependency-plugin/.config/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+ARG grafana_version=latest
```
why use latest and not a verified version?

Dhruv-J replied:
will update to 9.1.3

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/tsconfig.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "extends": "./.config/tsconfig.json"
+}
```
There are lots of generated file missing empty line. Do we have a method to avoid that? Or we should add it manually?

heanlan replied:
missing empty line

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/src/components/TablePanel.tsx:
```diff
@@ -0,0 +1,133 @@
+import React, { useMemo } from 'react';
```
Maybe we can add unit test for this file to verify if the data are in our component correctly?

Dhruv-J replied:
yes will do, working on getting tests to run locally

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/package.json:
```diff
@@ -0,0 +1,76 @@
+  "scripts": {
+    "build": "webpack -c ./.config/webpack/webpack.config.ts --env production",
+    "dev": "webpack -w -c ./.config/webpack/webpack.config.ts --env development",
+    "test": "jest --watch --onlyChanged",
```
Remove watch, otherwise it won't stop when we do the typescript unit test in the workflow.

Dhruv-J committed:
Application Layer Dashboard

This PR adds an Application Layer Dashboard to include visualizations for layer
seven Network Policies and HTTP values within the flows table.

Signed-off-by: Dhruv-J <dhruvj@vmware.com>

Dhruv-J committed:
linter fix dead links and addressed comments

Signed-off-by: Dhruv-J <dhruvj@vmware.com>

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/.config/.prettierrc.js:
```diff
@@ -0,0 +1,16 @@
+  "singleQuote": true,
+  "useTabs": false,
+  "tabWidth": 2
+};
```
empty line

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/.config/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+
+# Inject livereload script into grafana index.html
+USER root
+RUN sed -i 's/<\/body><\/html>/<script src=\"http:\/\/localhost:35729\/livereload.js\"><\/script><\/body><\/html>/g' /usr/share/grafana/public/views/index.html
```
ditto

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/.eslintrc:
```diff
@@ -0,0 +1,3 @@
+{
+  "extends": "./.config/.eslintrc"
+}
```
empty line

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/.gitignore:
```diff
@@ -0,0 +1,35 @@
+# Editor
+.idea
+
+.eslintcache
```
ditto

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/.nvmrc:
```diff
@@ -0,0 +1 @@
+16
```
ditto

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/.prettierrc.js:
```diff
@@ -0,0 +1,4 @@
+module.exports = {
+  // Prettier configuration provided by Grafana scaffolding
+  ...require("./.config/.prettierrc.js")
+};
```
ditto

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/package.json:
```diff
@@ -0,0 +1,79 @@
+    "typecheck": "tsc --noEmit",
+    "lint": "eslint --cache --ignore-path ./.gitignore --ext .js,.jsx,.ts,.tsx .",
+    "lint:fix": "npm run lint -- --fix",
+    "e2e": "npm exec cypress install && npm exec grafana-e2e run",
```
Remove unused scripts and dependencies, which might create unwanted security issue alert after it is merged.

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/src/components/TablePanel.test.tsx:
```diff
@@ -0,0 +1,68 @@
+    expect(renderedHtmlString.includes('GET')).toEqual(true);
+    expect(renderedHtmlString.includes('200')).toEqual(true);
+    expect(renderedHtmlString.includes('1256')).toEqual(true);
+    // expect(
```
Do you want to enable the below check? Or maybe remove it?

Dhruv-J replied:
yeah i'll remove it

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/src/components/TablePanel.test.tsx:
```diff
@@ -0,0 +1,68 @@
+    //   )
+    // ).toEqual(true);
+  });
+});
```
empty line

Dhruv-J committed:
fixes and feedback

Signed-off-by: Dhruv-J <dhruvj@vmware.com>

yuntanghsu commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/cypress/integration/01-smoke.spec.ts:
```diff
@@ -0,0 +1,10 @@
+import { e2e } from '@grafana/e2e';
+
+e2e.scenario({
```
Do we need this e2e test for Grafana? 
cc @heanlan

heanlan replied:
Can we run it with automation? e.g. will github action trigger it?

Dhruv-J replied:
I can remove this, I don't think it's testing anything important, nor is it run during yarn test.

Dhruv-J committed:
remove e2e default tests

Signed-off-by: Dhruv-J <dhruvj@vmware.com>

heanlan commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/src/components/TablePanel.tsx:
```diff
@@ -0,0 +1,136 @@
+      }
+    }
+    for (const txId in httpValsJSON) {
+      setTableRow(sourceIP+':'+sourcePort, destinationIP+':'+destinationPort, +txId, httpValsJSON[txId].hostname, httpValsJSON[txId].url, httpValsJSON[txId].http_user_agent, httpValsJSON[txId].http_content_type, httpValsJSON[txId].http_method, httpValsJSON[txId].protocol, httpValsJSON[txId].status, httpValsJSON[txId].length);
```
Why `+txId` instead of `txId`? Isn't `txId` already being a number here?

Dhruv-J replied:
since i'm interpreting the JSON, compiler thinks txID could have type string, and in order to convert it, I use the '+'

heanlan commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/src/components/TablePanel.test.tsx:
```diff
@@ -0,0 +1,46 @@
+    props.options = {};
+    let component = shallow(<TablePanel {...props} />);
+    React.useLayoutEffect = React.useEffect;
+    console.log('component text: '+component.render().text());
```
debug log?

heanlan commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/src/components/TablePanel.test.tsx:
```diff
@@ -0,0 +1,46 @@
+    let component = shallow(<TablePanel {...props} />);
+    React.useLayoutEffect = React.useEffect;
+    console.log('component text: '+component.render().text());
+    let renderedHtmlString = component.render().text();
```
I'm not familiar whether we can test it in another way, rather than converting to text. Maybe @yuntanghsu can give more constructive feedback.

But if we are just checking the text content, instead of using `includes` expression checking a series of substring ,can we just do `expect(textContent).toEqual(expectedContent);` to check the entire string?

yuntanghsu replied:
<img width="884" alt="image" src="https://github.com/antrea-io/theia/assets/59460118/03101fb0-c882-47f6-86cb-2d658195aede">
I think we should use `getAllByText` or `expect(getByText('10.10.1.4:42162')).toBeInTheDocument()` to check the data in the component.
For example, in your test and the picture, we can see there are 3 transaction IDs for a certain combination of source and destination. We want to make sure your component can generate exactly 3 elements, which should be based on the number of HttpVals you have. And we also want to make sure the contents are correct in these 3 elements. I'll suggest you change the vals in each transaction ID and verify the context for all of them. Otherwise the result can be a false positive case as you only use "includes" to check the data.


I also found there are lots of warning when using component.render().text():
```
at div
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at TableContainer (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/TableContainer/TableContainer.js:40:44)
        at MRT_TableContainer (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/table/MRT_TableContainer.tsx:13:38)
        at div
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at Paper (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/Paper/Paper.js:69:44)
        at MRT_TablePaper (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/table/MRT_TablePaper.tsx:11:34)
        at useRef (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/table/MRT_TableRoot.tsx:50:28)
        at MaterialReactTable (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/MaterialReactTable.tsx:14:3)
        at ThemeProvider (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/private-theming/node/ThemeProvider/ThemeProvider.js:39:5)
        at ThemeProvider (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/system/ThemeProvider/ThemeProvider.js:50:5)
        at ThemeProvider (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/styles/ThemeProvider.js:21:14)
        at div

      33 |     React.useLayoutEffect = React.useEffect;
      34 |     // console.log('component text: '+component.render().text());
    > 35 |     let renderedHtmlString = component.render().text();
         |                                        ^
      36 |     expect(renderedHtmlString.includes('10.10.1.4:42162')).toEqual(true);
      37 |     expect(renderedHtmlString.includes('93.184.216.34:80')).toEqual(true);
      38 |     expect(renderedHtmlString.includes('example.com')).toEqual(true);

      at printWarning (node_modules/react-dom/cjs/react-dom-server.node.development.js:82:30)
      at error (node_modules/react-dom/cjs/react-dom-server.node.development.js:58:5)
      at Object.useLayoutEffect (node_modules/react-dom/cjs/react-dom-server.node.development.js:1617:5)
      at useLayoutEffect (node_modules/react/cjs/react.development.js:1524:21)
      at useEventCallback (node_modules/@mui/utils/useEventCallback/useEventCallback.js:20:34)
      at useRippleHandler (node_modules/@mui/material/node/ButtonBase/ButtonBase.js:160:42)
      at Object.ButtonBase [as render] (node_modules/@mui/material/node/ButtonBase/ButtonBase.js:186:22)
      at ReactDOMServerRenderer.render (node_modules/react-dom/cjs/react-dom-server.node.development.js:3872:44)
      at ReactDOMServerRenderer.read (node_modules/react-dom/cjs/react-dom-server.node.development.js:3690:29)
      at Object.renderToStaticMarkup (node_modules/react-dom/cjs/react-dom-server.node.development.js:4314:27)
      at Object.renderToStaticMarkup [as render] (node_modules/enzyme-adapter-react-16/src/ReactSixteenAdapter.js:835:31)
      at ShallowWrapper.render (node_modules/enzyme/src/ShallowWrapper.js:1108:23)
      at ShallowWrapper.call [as single] (node_modules/enzyme/src/ShallowWrapper.js:1654:21)
      at ShallowWrapper.single [as html] (node_modules/enzyme/src/ShallowWrapper.js:1104:17)
      at ShallowWrapper.html [as render] (node_modules/enzyme/src/ShallowWrapper.js:1120:23)
      at Object.render (src/components/TablePanel.test.tsx:35:40)

  console.error
    Warning: useLayoutEffect does nothing on the server, because its effect cannot be encoded into the server renderer's output format. This will lead to a mismatch between the initial, non-hydrated UI and the intended UI. To avoid this, useLayoutEffect should only be used in components that render exclusively on the client. See https://reactjs.org/link/uselayouteffect-ssr for common fixes.
        at ButtonBase (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/ButtonBase/ButtonBase.js:95:44)
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at TableSortLabel (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/TableSortLabel/TableSortLabel.js:102:44)
        at span
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at Badge (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/Badge/Badge.js:169:44)
        at Tooltip (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/Tooltip/Tooltip.js:226:44)
        at MRT_TableHeadCellSortLabel (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/head/MRT_TableHeadCellSortLabel.tsx:14:3)
        at div
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at Box (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/system/createBox.js:31:41)
        at div
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at Box (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/system/createBox.js:31:41)
        at th
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at TableCell (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/TableCell/TableCell.js:108:44)
        at MRT_TableHeadCell (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/head/MRT_TableHeadCell.tsx:20:37)
        at tr
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at TableRow (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/TableRow/TableRow.js:70:44)
        at MRT_TableHeadRow (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/head/MRT_TableHeadRow.tsx:20:3)
        at thead
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at TableHead (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/TableHead/TableHead.js:44:44)
        at MRT_TableHead (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/head/MRT_TableHead.tsx:14:3)
        at table
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at Table (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/Table/Table.js:61:44)
        at MRT_Table (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/table/MRT_Table.tsx:19:29)
        at div
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at TableContainer (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/TableContainer/TableContainer.js:40:44)
        at MRT_TableContainer (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/table/MRT_TableContainer.tsx:13:38)
        at div
        at /Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23
        at Paper (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/Paper/Paper.js:69:44)
        at MRT_TablePaper (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/table/MRT_TablePaper.tsx:11:34)
        at useRef (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/table/MRT_TableRoot.tsx:50:28)
        at MaterialReactTable (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/material-react-table/src/MaterialReactTable.tsx:14:3)
        at ThemeProvider (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/private-theming/node/ThemeProvider/ThemeProvider.js:39:5)
        at ThemeProvider (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/system/ThemeProvider/ThemeProvider.js:50:5)
        at ThemeProvider (/Users/hsuy/Desktop/theia/plugins/grafana-custom-plugins/grafana-http-table-plugin/node_modules/@mui/material/node/styles/ThemeProvider.js:21:14)
        at div
```

@Dhruv-J Do you have any idea what happened?

Dhruv-J replied:
Yeah, the material-ui library [recommends](https://mui.com/material-ui/guides/testing/) running unit testing without the use of material-ui components in order to keep code immune to breaking due to changes within material-ui internal structure. So I can't directly check for including the table itself, so I check for the string being rendered within the html itself. @yuntanghsu I can attempt using those again, but I remember when I tried using similar solutions that false would be returned even though the text was supposedly being rendered. The warnings happen because the render() function is being called in a unit test, and so the result isn't actually visible anywhere. I looked into this issue specifically with material-ui and found that it is quite common, so people largely suppress the warning or haven't found a workaround. I'll try the clause that you mentioned and update the results within this thread.

heanlan commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/src/plugin.json:
```diff
@@ -0,0 +1,34 @@
+      "name": "Antrea io"
+    },
+    "logos": {
+      "small": "img/logo.svg",
```
please follow https://github.com/antrea-io/theia/blob/main/plugins/grafana-custom-plugins/grafana-chord-plugin/src/plugin.json#L14

heanlan replied:
not solved yet

heanlan commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/src/types.ts:
```diff
@@ -0,0 +1,7 @@
+type SeriesSize = 'sm' | 'md' | 'lg';
```
Are these options being used anywhere? I see they are passed as parameter into `TablePanel`, but I don't see where users can configure them, and where do we explicitly handle their values in the source code? The name "SimpleOptions" also looks a bit too general for me to understand its usage.

Dhruv-J replied:
those are default options, I'll remove them

heanlan reviewed (commented):
If we would like to deploy and test the dashboard by ourselves, how should we do it? Where can we get the sample http data? Could you give some instructions?

Dhruv-J commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/src/README.md:
will do, I'll replace the README in plugins/grafana-custom-plugins/grafana-http-table-plugin/ with the one in /src

Dhruv-J commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/src/components/TablePanel.tsx:
I'll move it to src/ since src/components is the autogenerated directory and other custom plugins contain code in src/ as well

Dhruv-J commented:
Instructions for those wanting to deploy the solution for themselves @heanlan:

1. pull Antrea PR [#5218](https://github.com/antrea-io/antrea/pull/5218)
2. pull Theia PR #417 and this PR
3. execute the included script for [build_antrea.sh.zip](https://github.com/antrea-io/theia/files/13639064/build_antrea.sh.zip)
4. deploy flow-visibility.yml from the new theia PR
5. apply the following: [pods.yaml.zip](https://github.com/antrea-io/theia/files/13639121/pods.yaml.zip)
6. annotate the pods with the following command: `kubectl pod annotate <pod-name> -n <namespace> [visibility.antrea.io/enable-l7=both](http://visibility.antrea.io/enable-l7=both)`
7. kubectl exec into the admin pod and run the command `curl example.com` to get one row, or `curl example.com example.com example.com` to get multiple rows (it'll take about 2 minutes to populate clickhouse)
8. open the application layer dashboard to view http values

Dhruv-J committed:
feedback addressed, unit tests and documentation fix

Signed-off-by: Dhruv-J <dhruvj@vmware.com>

Dhruv-J committed:
test and panel file updated

Signed-off-by: Dhruv-J <dhruvj@vmware.com>

heanlan commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/README.md:
```diff
@@ -1,114 +1,114 @@
+Supported Databases:
 
-1. Install dependencies
+- Clickhouse
```
ClickHouse

heanlan commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/src/plugin.json:
```diff
@@ -0,0 +1,34 @@
+    },
+    "logos": {
+      "small": "img/logo.svg",
+      "large": "img/logo.svg"
```
ditto

heanlan commented on plugins/grafana-custom-plugins/grafana-http-table-plugin/src/types.ts:
```diff
@@ -0,0 +1,2 @@
+export interface SimpleOptions {
```
rename to `TableOptions`?

heanlan reviewed (commented):
LGTM overall

Same with #488 , let's hold this PR up and take another look until all the PRs it depends on are merged.

Dhruv-J committed:
final feedback addressed

Signed-off-by: Dhruv-J <dhruvj@vmware.com>
